app/src/main/java/co/touchlab/kampkit/android/ui/Composables.kt

```diff
@@ -0,0 +1,155 @@
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.unit.dp
+import co.touchlab.kampkit.android.R
+import co.touchlab.kampkit.db.Breed
+import co.touchlab.kampkit.models.DataState
+import co.touchlab.kampkit.models.ItemDataSummary
+
+@Composable
+fun DogList(breeds: List<Breed>, onItemClick: (Breed) -> Unit) {
+    LazyColumn {
+        items(breeds) { breed ->
+            DogRow(breed) {
+                onItemClick(it)
+            }
+            Divider()
+        }
+    }
+}
+
+@Composable
+fun DogRow(breed: Breed, onClick: (Breed) -> Unit) {
+    Row(
+        Modifier
+            .clickable { onClick(breed) }
+            .padding(10.dp)
+    ) {
+        Text(breed.name, Modifier.weight(1F))
+        FavoriteIcon(breed)
+    }
+}
+
+@Composable
+fun FavoriteIcon(breed: Breed) {
+    Crossfade(
+        targetState = breed.favorite == 0L,
+        animationSpec = TweenSpec(
+            durationMillis = 500,
+            easing = FastOutSlowInEasing
+        )
+    ) { fav ->
+        if (fav) {
+            Image(
+                painter = painterResource(id = R.drawable.ic_favorite_border_24px),
+                contentDescription = "Favorite $breed"
+            )
+        } else {
+            Image(
+                painter = painterResource(id = R.drawable.ic_favorite_24px),
+                contentDescription = "Unfavorite $breed"
+            )
+        }
+    }
+}
+
+@Composable
+fun Loading(
+    lastState: DataState<ItemDataSummary>,
+    favoriteBreed: (Breed) -> Unit
+) {
+    when (lastState) {
+        DataState.Loading -> {
+            // Nothing here, because it's taken care of by SwipeRefresh
+        }
+        DataState.Empty -> {
+            Empty()
+        }
+        is DataState.Error -> {
+            Error(errorState = lastState)
+        }
+        is DataState.Success -> {
+            Success(lastState, favoriteBreed)
+        }
+    }
+}
+
+@Composable
+fun Empty() {
+    Column(
+        modifier = Modifier.fillMaxSize()
+            .verticalScroll(rememberScrollState()),
+        verticalArrangement = Arrangement.Center,
+        horizontalAlignment = Alignment.CenterHorizontally,
+    ) {
+        Text("Sorry, no doggos found")
+    }
+}
+
+@Composable
+fun Error(errorState: DataState.Error) {
+    Column(
+        modifier = Modifier.fillMaxSize()
+            .verticalScroll(rememberScrollState()),
+        verticalArrangement = Arrangement.Center,
+        horizontalAlignment = Alignment.CenterHorizontally,
+    ) {
+        Text(text = errorState.exception)
+    }
+}
+
+@Composable
+fun Success(
+    successState: DataState.Success<ItemDataSummary>,
+    favoriteBreed: (Breed) -> Unit
+) {
+    DogList(breeds = successState.data.allItems, favoriteBreed)
+}
+
+@Composable
+fun MainScreen(
+    dataState: DataState<ItemDataSummary>,
+    favoriteBreed: (Breed) -> Unit
+) {
+    val lastState: MutableState<DataState<ItemDataSummary>> = remember { mutableStateOf(dataState) }
```

Comment:
I'm doing this weird `lastState` business, because of an issue with all the content going away in the `Loading` state. My first approach was:
``` kotlin
when (dataState) {
    DataState.Loading -> {
        // Nothing, taken care of by SwipeFresh
    }
    DataState.Empty -> {
        Empty()
    }
    is DataState.Error -> {
        Error(dataState.exception)
    }
    is DataState.Success<ItemDataSummary> -> {
        Success(favoriteBreed)
    }
}
```
And this is a problem because if no UI is emitted in the `Loading` state, then the screen is blank. What I wanted to do in the `Loading` state is still show whatever was the last thing we showed, and show the refresh over it. That's why the `Loading` Composable is doing more work than one would think. Was thinking of removing `Loading` as a valid `DataState`, and making `Loading` its own `StateFlow<Boolean>`, but that would require updating iOS as well, and I wanted to make this first PR the least disruptive as possible.

Reply:
Ahh that makes sense now. The examples of swiperefresh do show a separate loading boolean field on their data class rather than a Loading entry in the sealed class since to keep the old data on screen compose needs access to it in the loading state. Adding a `data: T?` to `DataState.Loading` might be another solution that would be less change on ios side.

Reply:
Yeah, that might be a good way to do it as well. `StateFlow` doesn't allow holding more than 1 value, but `SharedFlow` can be configured to keep the last n-items it emitted in a replay buffer, like this:

```kotlin
val shared = MutableSharedFlow(
    replay = 2,
    onBufferOverflow = BufferOverflow.DROP_OLDEST
)
```
 Then the data structure can stay the same, and iOS doesn't need to worry about changing its implementation. That might be more flexible for swapping View / Compose, or UI Kit / Swift UI. Will tinker with a few ideas and report back.

Reply:
I *think* there's a flow operator that'll transform emissions into a pair of the previous value and new value which might be helpful, but I can't remember what its called. Although I think for a more greenfield situation the simple solution of changing the data type thats emitted is probably the way to go

Reply:
There's an operator called `buffer()`, which seems like it would accomplish this, but it doesn't actually make the buffer available to the caller. It just allows collectors to operate on a value immediately after getting it, so slow producers don't block collectors. More [here](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/buffer.html)